### PR TITLE
Introduce xml subtypes

### LINKED ISSUE #526: Implement xml subtypes in types module
The types module supports the `xml` type. This adds support for xml subtypes.

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
Add xml:Element, xml:Comment, xml:ProcessingInstruction, xml:Text, xml&lt;xml:Text&gt;, and so on.

Fixes https://github.com/ballerina-platform/nballerina/issues/526

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/types/xml.bal [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        rw: {
+            data: rwBitset,
+            sequence: false
+        }
+    };
+    UniformTypeCode typeCode = rwBitset == 0 ? UT_XML_RO : UT_XML_RW;
+    return uniformSubtype(typeCode, st);
+}
+
+public function xmlSequence(SemType constituentType) returns SemType {
```
Comment (jclark) [inline compiler/modules/types/xml.bal:37]: Need to be clear about how this function deals with case when constituent type is not a subtype of XML.

File: compiler/modules/types/xml.bal [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    if constituentType == XML {
+        return constituentType;
+    }
+    var roData = getComplexSubtypeData(<ComplexSemType>constituentType, UT_XML_RO);
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: These casts are not safe.

File: compiler/modules/types/xml.bal [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    return { ro, rw };
+}
+
+function xmlSubtypeUnion(SubtypeData d1, SubtypeData d2) returns SubtypeData {
+    XmlSubtype v1 = <XmlSubtype>d1;
+    XmlSubtype v2 = <XmlSubtype>d2;
+    int dataRo = v1.ro.data | v2.ro.data;
+    int dataRw = v1.rw.data | v2.rw.data;
+    return createXmlSubtype(
+        { data :dataRo, sequence: bddUnion(v1.ro.sequence, v2.ro.sequence) },
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: Space like `data: dataRo`.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/types/xml.bal [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    else {
+        return d;
+    }
+}
+
+function createXmlSubtype(int data, Bdd sequence) returns SubtypeData {
+    if data == 0 && sequence == false {
+        return false;
+    }
+    else if (data & XML_NEVER) != 0 && sequence == true {
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: This isn't right: you need to check that it allows everything that is possible, which is different for the ro/rw case, so you will need an argument to the function.

File: compiler/modules/types/xml.bal [hunk 5]
```diff
@@ -0,0 +1,188 @@
+// Implementation specific to basic type xml.
+import wso2/nballerina.err;
+
+public type XmlSubtypeData readonly & record {|
+    int data;
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: `data` isn't a good name.  This needs a comment explaining what the values are.

File: compiler/modules/types/xml.bal [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    Bdd sequence;
+|};
+
+const int XML_NEVER = 1;
+const int TEXT = 1 << 1;
+const int ELEM_RO = 1 << 2;
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: ELEMENT is more consistent

File: compiler/modules/types/xml.bal [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+const int PI_RW = 1 << 6;
+const int COMMENT_RW = 1 << 7;
+
+const int XML_RO_MASK = XML_NEVER | TEXT | ELEM_RO | PI_RO | COMMENT_RO;
+const int XML_RW_MASK = ELEM_RW | PI_RW | COMMENT_RW;
+const int XML_RO_SEQ_MASK = TEXT | ELEM_RO | PI_RO | COMMENT_RO;
+
+final XmlSubtypeData emptyXmlSubtype = { data: 0, sequence: false };
+
+function xmlSingleton(int bits) returns SemType {
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: You are using `bits` here to mean what `data` means in XmlSubtypeData.

File: compiler/modules/types/xml.bal [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        { data: bits & XML_RW_MASK, sequence: false }
+    );
+}
+
+public function xmlSequence(SemType constituentType, err:Location loc) returns SemType|err:Semantic {
+    if constituentType == XML {
+        return constituentType;
+    }
+    if constituentType is UniformTypeBitSet {
+        return err:semantic("Invalid xml constituent type", loc=loc);
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: Don't capitalize error messages,

File: compiler/modules/types/xml.bal [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+    }
+    ComplexSemType t = <ComplexSemType>constituentType;
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: Use `else` rather than casting.

File: compiler/modules/types/xml.bal [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    if hasNonXmlTypes(t) {
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: Just start off at the beginning by doing if !t:isSubtypeSimple(ty, t:XML)`

File: compiler/modules/types/xml.bal [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+    return (t.all & ~XML) != 0 || (t.some & ~XML) != 0;
+}
+
+
+function makeSequence(XmlSubtypeData d) returns XmlSubtypeData {
+    if d.sequence == false {
+        return  { data: XML_NEVER | d.data, sequence: bddCreate(d.data, true, false, false) };
+    }
+    else {
+        return d;
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: That's not right. The singletons can occur in the sequence

File: compiler/modules/types/xml.bal [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+    }
+}
+
+function createXmlSubtype(boolean isRo, int data, Bdd sequence) returns SubtypeData {
+    if data == 0 && sequence == false {
+        return false;
+    }
+    boolean containsAllData = isRo ? (data & XML_RO_MASK) == XML_RO_MASK : (data & XML_RW_MASK) == XML_RW_MASK;
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: I would write it like this:

```
if sequence == false {
    if data == 0 {
        return false;
    }
}
else if sequence == true {
    int mask = isRo ? XML_RO_MASK : XML_RW_MASK;
    if (data & mask) == mask {
        return true;
    }
}
return { data, sequence };
```

File: compiler/modules/types/xml.bal [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+    if containsAllData && sequence == true {
+        return true;
+    }
+    return { data, sequence };
+}
+
+function createXmlSemtype(XmlSubtypeData ro, XmlSubtypeData rw) returns ComplexSemType {
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: This function doesn't seem right to me. Why are we getting an emptyXmlSubtype?

File: compiler/modules/types/xml.bal [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+function xmlSubtypeDiff(boolean isRo, SubtypeData d1, SubtypeData d2) returns SubtypeData {
+    XmlSubtypeData v1 = <XmlSubtypeData>d1;
+    XmlSubtypeData v2 = <XmlSubtypeData>d2;
+    int data = v1.data & ~v2.data;
+    return createXmlSubtype(isRo, data, bddDiff(v1.sequence, v2.sequence));
+}
+
+function xmlSubtypeComplement(boolean isRo, SubtypeData d) returns SubtypeData {
```
Comment (jclark) [inline compiler/modules/types/xml.bal:112]: This doesn't look right.

File: compiler/modules/types/xml.bal [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+function xmlSubtypeUnion(boolean isRo, SubtypeData d1, SubtypeData d2) returns SubtypeData {
+    XmlSubtypeData v1 = <XmlSubtypeData>d1;
+    XmlSubtypeData v2 = <XmlSubtypeData>d2;
+    int data = v1.data | v2.data;
+    return createXmlSubtype(isRo, data, bddUnion(v1.sequence, v2.sequence));
+}
+
+function xmlSubtypeIntersect(boolean isRo, SubtypeData d1, SubtypeData d2) returns SubtypeData {
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: intersect of two sets can only produce something smaller than the args. So there is no possibility of producing true from this. So you don't need isRo as an argument (because isRo is only used to detect whether you've gort true.)  So the ops tables can use this directly. Similarly for diff.

File: compiler/modules/types/xml.bal [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+    Bdd seq = bddCreate(rootAtom, bddComplement(sd.sequence), false, false);
+    return createXmlSubtype(isRo, complement, seq);
+}
+
+function xmlSubtypeIsEmpty(Context cx, SubtypeData d) returns boolean {
+    XmlSubtypeData sd = <XmlSubtypeData>d;
+    if sd.data != 0 {
+        return false;
+    }
+    return !xmlSequenceInhabited(cx, sd.sequence);
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: I would call this xmlBddEmpty and have it return the negation of what or currently returns

File: compiler/modules/types/xml.bal [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+function xmlSequenceInhabited(Context cx, Bdd bdd) returns boolean {
+    boolean isEmpty = bddEvery(cx, bdd, (), (), xmlFormulaIsEmpty);
+    return !isEmpty;
+}
+
+function xmlFormulaIsEmpty(Context cx, Conjunction? pos, Conjunction? neg) returns boolean {
+    return hasIntersectionAll(allBits(pos), neg);
+}
+
+function hasIntersectionAll(int bits, Conjunction? c) returns boolean {
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: I don't think it's helpful to write this function or `allBits` recursively.

xmlFormulaIsEmpty should just loop over pos to build an `int` with the bits.
Then loop over the neg, returing true if any neg has all the pos bits.

**Review by jclark — COMMENTED** [review]
This looks like it is on the right lines.  I haven't looked at the other files yet.

File: compiler/modules/types/xml.bal (hunk 15 above)
Comment (rdhananjaya) [reply compiler/modules/types/xml.bal]: So inline the call to createXmlSubtype?

Comment (jclark) [reply compiler/modules/types/xml.bal]: Factor out the part of createXmlSubtype that deals with false.

File: compiler/modules/front/resolveTypes.bal [hunk 18]
```diff
@@ -229,6 +229,10 @@ function resolveTypeDesc(ModuleSymbols mod, s:ModuleLevelDefn modDefn, int depth
     if td is s:XmlSequenceTypeDesc {
         t:SemType t = check resolveTypeDesc(mod, modDefn, depth, td.constituent);
         err:Location loc =  err:location(modDefn.part.file, td.pos);
+        
+        if !t:isSubtypeSimple(t, t:XML) {
+            return err:semantic("non xml type as xml constituent type", loc=loc);
```
Comment (manuranga) [inline compiler/modules/front/resolveTypes.bal]: Should this be `constraint type`?

Comment (manuranga) [reply compiler/modules/front/resolveTypes.bal]: Same for the `XmlSequenceTypeDesc` field.

File: compiler/modules/types/xml.bal [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+|};
+
+const int XML_NEVER = 1;
+const int TEXT = 1 << 1;
+const int ELEMENT_RO = 1 << 2;
+const int PI_RO = 1 << 3;
+const int COMMENT_RO = 1 << 4;
+const int ELEMENT_RW = 1 << 5;
+const int PI_RW = 1 << 6;
+const int COMMENT_RW = 1 << 7;
```
Comment (manuranga) [inline compiler/modules/types/xml.bal]: minor: consider aligning.

File: compiler/modules/types/xml.bal [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+    }
+    return createXmlSubtypeOrEmpty(primitives, sequence);
+}
+
+function createXmlSubtypeOrEmpty(int primitives, Bdd sequence) returns SubtypeData {
+    if sequence == false {
+        if primitives == 0 {
+            return false;
+        }
+    }
```
Comment (manuranga) [inline compiler/modules/types/xml.bal]: Can combine the `if`s.

File: compiler/modules/types/xml.bal [hunk 21]
```diff
@@ -0,0 +1,164 @@
...
+import wso2/nballerina.err;
+
+public type XmlSubtypeData readonly & record {|
+    // Each xml singleton and the empty sequence is considered a primitive.
+    int primitives;
+    Bdd sequence;
+|};
+
+const int XML_NEVER = 1;
+const int TEXT = 1 << 1;
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: We should put XML_ before these since their scope is the entire module not just this file.

File: compiler/modules/types/xml.bal [hunk 22]
```diff
@@ -0,0 +1,164 @@
+// Implementation specific to basic type xml.
+import wso2/nballerina.err;
+
+public type XmlSubtypeData readonly & record {|
+    // Each xml singleton and the empty sequence is considered a primitive.
+    int primitives;
+    Bdd sequence;
+|};
+
+const int XML_NEVER = 1;
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: There needs to be a comment that connects these consts to the primitives field and explains that the primitives field is a bitwise-or of these constants.

File: compiler/modules/types/xml.bal [hunk 23]
```diff
@@ -0,0 +1,164 @@
...
+const int COMMENT_RO = 1 << 4;
+const int ELEMENT_RW = 1 << 5;
+const int PI_RW = 1 << 6;
+const int COMMENT_RW = 1 << 7;
+
+const int XML_RO_MASK = XML_NEVER | TEXT | ELEMENT_RO | PI_RO | COMMENT_RO;
+const int XML_RW_MASK = ELEMENT_RW | PI_RW | COMMENT_RW;
+
+final XmlSubtypeData xmlRoAll = { primitives: XML_RO_MASK, sequence: bddAtom(XML_RO_MASK & ~XML_NEVER) };
+final XmlSubtypeData xmlRwAll = { primitives: XML_RW_MASK, sequence: bddAtom(XML_RW_MASK) };
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: You can have a XML value that is a sequence of length >= 2 that includes both mutable and immutable singletons, for example one mutable XML element singleton and one immutable XML element singleton. Such a sequence would be part of the XML_RW uniform type. This means that bddAtom in the Rw case needs to use a mask that allows immutable things as well as mutable things.

File: compiler/modules/types/xml.bal [hunk 24]
```diff
@@ -0,0 +1,164 @@
...
+const int XML_NEVER = 1;
+const int TEXT = 1 << 1;
+const int ELEMENT_RO = 1 << 2;
+const int PI_RO = 1 << 3;
+const int COMMENT_RO = 1 << 4;
+const int ELEMENT_RW = 1 << 5;
+const int PI_RW = 1 << 6;
+const int COMMENT_RW = 1 << 7;
+
+const int XML_RO_MASK = XML_NEVER | TEXT | ELEMENT_RO | PI_RO | COMMENT_RO;
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: I would define XML_RO_SINGLETONS (excluding XML_NEVER) and then use that later on.

File: compiler/modules/types/xml.bal [hunk 25]
```diff
@@ -0,0 +1,164 @@
...
+const int XML_RW_MASK = ELEMENT_RW | PI_RW | COMMENT_RW;
+
+final XmlSubtypeData xmlRoAll = { primitives: XML_RO_MASK, sequence: bddAtom(XML_RO_MASK & ~XML_NEVER) };
+final XmlSubtypeData xmlRwAll = { primitives: XML_RW_MASK, sequence: bddAtom(XML_RW_MASK) };
+
+function xmlSingleton(int primitives) returns SemType {
+    return createXmlSemtype(
+        { primitives: primitives & XML_RO_MASK, sequence: false },
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: I believe you need to use createXmlSubtype before calling createXmlSemtype.

File: compiler/modules/types/xml.bal [hunk 26]
```diff
@@ -0,0 +1,164 @@
...
+final XmlSubtypeData xmlRwAll = { primitives: XML_RW_MASK, sequence: bddAtom(XML_RW_MASK) };
+
+function xmlSingleton(int primitives) returns SemType {
+    return createXmlSemtype(
+        { primitives: primitives & XML_RO_MASK, sequence: false }, 
+        { primitives: primitives & XML_RW_MASK, sequence: false }
+    );
+}
+
+public function xmlSequence(SemType constituentType, err:Location loc) returns SemType {
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: Consider `xml<xml:Element>`. The XML_RW half of this has to allow a combination of mutable and immutable elements. In other words, you cannot do makeSequence independently for the ro/rw cases.

File: compiler/modules/types/xml.bal [hunk 27]
```diff
@@ -0,0 +1,164 @@
...
+        return false;
+    }
+    return xmlBddEmpty(cx, sd.sequence);
+}
+
+function xmlBddEmpty(Context cx, Bdd bdd) returns boolean {
+    return bddEvery(cx, bdd, (), (), xmlFormulaIsEmpty);
+}
+
+function xmlFormulaIsEmpty(Context cx, Conjunction? pos, Conjunction? neg) returns boolean {
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: For the RW case, we also have to detect when it allows only RO constituents.

File: compiler/modules/types/xml.bal [hunk 28]
```diff
@@ -0,0 +1,164 @@
...
+public function xmlSequence(SemType constituentType, err:Location loc) returns SemType {
+    if constituentType is UniformTypeBitSet {
+        return constituentType;
+    }   
+    else {
+        SubtypeData ro = getComplexSubtypeData(constituentType, UT_XML_RO);
+        ro = ro is boolean ? ro : makeSequence(<XmlSubtypeData>ro);
+        
+        SubtypeData rw = getComplexSubtypeData(constituentType, UT_XML_RW);
+        rw = rw is boolean ? rw : makeSequence(<XmlSubtypeData>rw);
```
Comment (manuranga) [inline compiler/modules/types/xml.bal]: Shouldn't some reusing of BDDs happen here? or is it happening elsewhere?

**Review by jclark — CHANGES_REQUESTED** [review]
I think we're nearly there, but we have to deal with sequences that contain a mixture of mutable and immutable singletons.

File: compiler/modules/types/xml.bal [hunk 29]
```diff
@@ -0,0 +1,164 @@
...
+        
+        return createXmlSemtype(ro, rw);
+    }
+}
+
+function makeSequence(XmlSubtypeData d) returns XmlSubtypeData {
+    return { 
+        primitives: XML_NEVER | d.primitives, 
+        sequence: bddUnion(bddAtom(d.primitives & ~XML_NEVER), d.sequence) 
+        };
```
Comment (manuranga) [inline compiler/modules/types/xml.bal]: minor: formatting looks bit off to me.

File: compiler/modules/front/resolveTypes.bal (hunk 18 above)
Comment (rdhananjaya) [reply compiler/modules/front/resolveTypes.bal]: `constituent` is spec term used in XML sequences.

File: compiler/modules/types/xml.bal (hunk 21 above)
Comment (rdhananjaya) [reply compiler/modules/types/xml.bal]: How about `XML_TEXT_BIT`, `XML_TEXT` is already taken for the semtype.

File: compiler/modules/front/resolveTypes.bal (hunk 18 above)
Comment (manuranga) [reply compiler/modules/front/resolveTypes.bal]: `constituent` in spec refer to the values not the type `T` in `xml<T>`.
But it makes sense here, because it's followed by the word 'type', not sure about it's use in `XmlSequenceTypeDesc`.

File: compiler/modules/types/xml.bal (hunk 21 above)
Comment (jclark) [reply compiler/modules/types/xml.bal]: I suggest you use a prefix of `XML_PRIMITIVE_` for all of them, which ties them to the `primitives` field.

**Comment by jclark** [comment]
I just merged in Ushira's new semtype testing stuff. Shall we try to use that? The readonly aspect of this is particularly tricky, and we are inventing all this rather than relying on the Castagna paper, so we need tests that probe all the difficult edge cases.

File: compiler/modules/front/resolveTypes.bal (hunk 18 above)
Comment (rdhananjaya) [reply compiler/modules/front/resolveTypes.bal]: I really don't have a strong opinion on this. I was tempted to use `constraint` here as it's what we have in jBallerina type. But since xml sequence is immutable, I didn't feel constraint means anything here. In a map<T>, T is a constraint as we can't manipulate the value to go over T, but in XML case it just who what is this sequence made out of.

File: compiler/modules/front.syntax/parseTypeDesc.bal [hunk 30]
```diff
@@ -94,7 +93,6 @@ function parsePrimaryTypeDesc(Tokenizer tok) returns TypeDesc|err:Syntax {
         |  "readonly" => {
             Position endPos = tok.currentEndPos();
             check tok.advance();
-            // JBUG should not need cast #30191
```
Comment (manuranga) [inline compiler/modules/front.syntax/parseTypeDesc.bal]: is this removed by accident ?

Comment (rdhananjaya) [reply compiler/modules/front.syntax/parseTypeDesc.bal]: oh, yes my bad, let me reset this

File: compiler/modules/types/xml.bal [hunk 31]
```diff
@@ -0,0 +1,196 @@
...
+const int XML_PRIMITIVE_NEVER_BIT      = 1;
+const int XML_PRIMITIVE_TEXT_BIT       = 1 << 1;
+const int XML_PRIMITIVE_ELEMENT_RO_BIT = 1 << 2;
+const int XML_PRIMITIVE_PI_RO_BIT      = 1 << 3;
+const int XML_PRIMITIVE_COMMENT_RO_BIT = 1 << 4;
+const int XML_PRIMITIVE_ELEMENT_RW_BIT = 1 << 5;
+const int XML_PRIMITIVE_PI_RW_BIT      = 1 << 6;
+const int XML_PRIMITIVE_COMMENT_RW_BIT = 1 << 7;
+
+const int XML_PRIMITIVE_RO_SINGLTONS = XML_PRIMITIVE_TEXT_BIT | XML_PRIMITIVE_ELEMENT_RO_BIT | XML_PRIMITIVE_PI_RO_BIT | XML_PRIMITIVE_COMMENT_RO_BIT;
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: SINGLETONS has an E

File: compiler/modules/types/core.bal [hunk 32]
```diff
@@ -188,7 +188,7 @@ public class Context {
     }
 }
 
-type ProperSubtypeData StringSubtype|DecimalSubtype|FloatSubtype|IntSubtype|BooleanSubtype|BddNode;
+type ProperSubtypeData StringSubtype|DecimalSubtype|FloatSubtype|IntSubtype|BooleanSubtype|XmlSubtypeData|BddNode;
```
Comment (jclark) [inline compiler/modules/types/core.bal]: Naming here is not consistent. Should use `XmlSubtype` here.

File: compiler/modules/types/core.bal [hunk 33]
```diff
@@ -364,6 +364,10 @@ public final UniformTypeBitSet SIMPLE_OR_STRING = uniformTypeUnion((1 << UT_NIL)
 public final UniformTypeBitSet NUMBER = uniformTypeUnion((1 << UT_INT) | (1 << UT_FLOAT) | (1 << UT_DECIMAL));
 public final SemType BYTE = intWidthUnsigned(8);
 public final SemType STRING_CHAR = stringChar();
+public final SemType XML_PRIMITIVE_ELEMENT = xmlSingleton(XML_PRIMITIVE_ELEMENT_RO_BIT | XML_PRIMITIVE_ELEMENT_RW_BIT);
```
Comment (jclark) [inline compiler/modules/types/core.bal]: No need for `_PRIMITIVE` in the name of these.

File: compiler/modules/types/xml.bal [hunk 34]
```diff
@@ -0,0 +1,196 @@
...
+    // Each xml singleton and the empty sequence is considered a primitive.
+    int primitives;
+    Bdd sequence;
+|};
+
+// Each bit represents a XML singleton subtype or a empty sequence.
+// `primitives` fields are composed of allowed values from this singleton set.
+// Atom of the `Bdd` represented by `sequence` field is also composed of these to indicate 
+// non-empty sequence of allowed singletons.
+const int XML_PRIMITIVE_NEVER_BIT      = 1;
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: I don't think we need BIT in the name of these as well as PRIMITIVE.

File: compiler/modules/types/xml.bal [hunk 35]
```diff
@@ -0,0 +1,196 @@
+// Implementation specific to basic type xml.
+
+public type XmlSubtypeData readonly & record {|
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: This should be just XmlSubtype.

File: compiler/modules/front/resolveTypes.bal (hunk 18 above)
Comment (jclark) [reply compiler/modules/front/resolveTypes.bal]: "type parameter for xml is not a subtype of xml"

File: compiler/modules/front/resolveTypes.bal [hunk 36]
```diff
@@ -252,6 +252,15 @@ function resolveTypeDesc(ModuleSymbols mod, s:ModuleLevelDefn modDefn, int depth
             return t:intersect(l, r);
         }
     }
+    if td is s:XmlSequenceTypeDesc {
+        t:SemType t = check resolveTypeDesc(mod, modDefn, depth, td.constituent);
+        err:Location loc =  err:location(modDefn.part.file, td.pos);
+        
+        if !t:isSubtypeSimple(t, t:NEVER) && !t:isSubtypeSimple(t, t:XML) {
```
Comment (jclark) [inline compiler/modules/front/resolveTypes.bal]: `!t:isSubtypeSimple(t, t:NEVER)` is the same as `t != t:NEVER`. But why do we need this? IIRC, `isSubtypeSimple(t:NEVER, t:XML)` is true.

**Review by jclark — COMMENTED** [review]
Are we recognizing that `xml<xml:Text>` and `xml:Text` are the same thing? I don't think we are...

I also think we need to be able to test with `!` to make sure we get stuff right. We can now do that: https://github.com/ballerina-platform/nballerina/issues/551#issuecomment-953462130

**Comment by jclark** [comment]
I have now implemented  a `!` type operator that is allowed in type descriptors when you are doing showTypes or `-t` tests. So we can now put in tests for all the various difficult cases we discussed.

File: compiler/testSuite/semtype/xml-complex-ro-t.bal [hunk 37]
```diff
@@ -0,0 +1,41 @@
...
+// @type E < NonEmptyS
+// @type P < NonEmptyS
+// @type C < NonEmptyS
+type NonEmptyS S & !N;
+
+// @type NonEmptyS < UX
+type UX XE|XP|XC|T;
+
+// @type XNonEmptyS = RX
+// @type XNonEmptyS < X
```
Comment (jclark) [inline compiler/testSuite/semtype/xml-complex-ro-t.bal]: Don't understand this one: `xml<T>` means sequence of length 0 or more where every member belongs to T. So I would say: XNonEmptyS = X. What am I missing?

**Review by jclark — COMMENTED** [review]
I still need to look at this some more.

File: compiler/modules/types/xml.bal [hunk 38]
```diff
@@ -0,0 +1,196 @@
...
+        
+        SubtypeData rw = getComplexSubtypeData(constituentType, UT_XML_RW);
+        rw = rw is boolean ? rw : makeSequence(false, <XmlSubtype>rw);
+        
+        return createXmlSemtype(ro, rw);
+    }
+}
+
+function makeSequence(boolean roPart, XmlSubtype d) returns XmlSubtype {
+    return {
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: Shouldn't this use createXmlSubtype? Otherwise, you are not going to recognize all the cases where this is equivalent to true/false.

File: compiler/modules/types/xml.bal [hunk 39]
```diff
@@ -0,0 +1,196 @@
...
+        ro = ro is boolean ? ro : makeSequence(true, <XmlSubtype>ro);
+        
+        SubtypeData rw = getComplexSubtypeData(constituentType, UT_XML_RW);
+        rw = rw is boolean ? rw : makeSequence(false, <XmlSubtype>rw);
+        
+        return createXmlSemtype(ro, rw);
+    }
+}
+
+function makeSequence(boolean roPart, XmlSubtype d) returns XmlSubtype {
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: How does this deal with the case where we are creating the rw part, and d.sequence is false and all the primitives are ro?

e.g. `xml<readonly & xml:Element> & !readonly` should equivalent to `never`.

Comment (rdhananjaya) [reply compiler/modules/types/xml.bal]: I think I forgot to update this when addressing https://github.com/ballerina-platform/nballerina/pull/562#pullrequestreview-780698845; will update.

File: compiler/testSuite/semtype/xml-complex-ro-t.bal (hunk 37 above)
Comment (rdhananjaya) [reply compiler/testSuite/semtype/xml-complex-ro-t.bal]: I improved the names reflect the type.
Here since T is `readonly`, xml<T> can only contain readonly parts hence it's a subtype of `X`

Comment (jclark) [reply compiler/testSuite/semtype/xml-complex-ro-t.bal]: OK. Got it.

File: compiler/modules/types/xml.bal [hunk 40]
```diff
@@ -0,0 +1,197 @@
...
+const int XML_PRIMITIVE_PI_RO      = 1 << 3;
+const int XML_PRIMITIVE_COMMENT_RO = 1 << 4;
+const int XML_PRIMITIVE_ELEMENT_RW = 1 << 5;
+const int XML_PRIMITIVE_PI_RW      = 1 << 6;
+const int XML_PRIMITIVE_COMMENT_RW = 1 << 7;
+
+const int XML_PRIMITIVE_RO_SINGLETONS = XML_PRIMITIVE_TEXT | XML_PRIMITIVE_ELEMENT_RO | XML_PRIMITIVE_PI_RO | XML_PRIMITIVE_COMMENT_RO;
+const int XML_PRIMITIVE_RO_MASK = XML_PRIMITIVE_NEVER | XML_PRIMITIVE_RO_SINGLETONS;
+const int XML_PRIMITIVE_RW_MASK = XML_PRIMITIVE_ELEMENT_RW | XML_PRIMITIVE_PI_RW | XML_PRIMITIVE_COMMENT_RW;
+const int XML_RW_ATOM = XML_PRIMITIVE_RO_SINGLETONS | XML_PRIMITIVE_RW_MASK;
```
Comment (jclark) [inline compiler/modules/types/xml.bal]: This name seems a bit inconsistent with the others. I would call this `XML_PRIMITIVE_SINGLETON` and rename `XML_PRIMITIVE_RO_SINGLETONS` to `XML_PRIMITIVE_RO_SINGLETON`.

**Review by jclark — CHANGES_REQUESTED** [review]
Looks just. Just a small naming tweak and we are done!

**Review by jclark — CHANGES_REQUESTED** [review]
Can you also update docs?

* Add to docs/type-subset.md
* Cross off from compiler/modules/types/TODO.md
* In compiler/modules/types/NOTES.md, update and move to the Done section
